[AutoDiff] Update @differentiable attribute, refactor `lookupFuncDecl` method.

dan-zheng opened the pull request:
- Add adjoint access level check to `@differentiable` attribute.
  - If the original function is "exported" (public or @_versioned), then the
    adjoint must also be exported.
- Generalize `lookupFuncDecl` function and add it as a method to `TypeChecker`
  so that it can be used for type-checking `#adjoint` expression in a later
  commit.

dan-zheng committed:
[AutoDiff] Update @differentiable attribute, refactor `lookupFuncDecl` method.

- Add adjoint access level check to @differentiable attribute.
  - If the original function is "exported" (public or @_versioned), then the
    adjoint must also be exported.
- Generalize `lookupFuncDecl` function and add it as a method to `TypeChecker`
  so that it can be used for type-checking #adjoint expression in a later
  commit.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2374,6 +2288,11 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
+    std::function<void()> primalInvalidTypeContextDiagnostic = [&]() {
+      TC.diagnose(primalNameLoc,
+                  diag::differentiable_attr_function_not_same_type_context,
+                  primalSpecifier.Name);
```
Do you need a `setInvalid` here?

dan-zheng replied:
I think `setInvalid` is handled by the block below:
```c++
primal = TC.lookupFuncDecl(...)
if (!primal) {
  attr->setInvalid();
  return;
}
```

If any diagnostic is emitted, `primal` will be `nullptr` and `attr->setInvalid()` will execute.

rxwei commented on lib/Sema/TypeCheckExpr.cpp:
```diff
@@ -709,3 +709,107 @@ bool TypeChecker::isCompatibleWithVectorAutoDiff(Type type, DeclContext *DC) {
+      notAFuncDecl = true;
+      continue;
+    }
+    if (hasValidTypeCtx.hasValue()) {
```
```cpp
if (hasValidTypeCtx && !(*hasValidTypeCtx)(funcDecl)) {
  wrongTypeContext = true;
  continue;
}
```

dan-zheng replied:
Done.

rxwei commented on lib/Sema/TypeCheckExpr.cpp:
```diff
@@ -709,3 +709,107 @@ bool TypeChecker::isCompatibleWithVectorAutoDiff(Type type, DeclContext *DC) {
+  // If function declaration could not be resolved, emit the appropriate
+  // diagnostic.
+  if (!resolvedFuncDecl) {
+    if (results.empty()) {
```
Instead of these `if-else`s, I would suggest 
```cpp
if (results.empty()) {
  diagnose(...)
  return;
}

if (ambiguousFuncDecl) {
  ambiguousDiagnostic();
  return;
}

...
```

dan-zheng replied:
Done.

rxwei commented on lib/Sema/TypeCheckExpr.cpp:
```diff
@@ -709,3 +709,107 @@ bool TypeChecker::isCompatibleWithVectorAutoDiff(Type type, DeclContext *DC) {
+  // the type context.
+  // This works around the fact that operators cannot be specified with a
+  // qualified name (i.e. only `(+)` works, `Float.(+)` doesn't).
+  if (funcName.isOperator() && lookupContext->isTypeContext()) {
```
How about

```cpp
if (funcName.isOperator() && lookupContext->isTypeContext())
  if (auto tmp = lookupMember(lookupContext, lookupContext->getSelfTypeInContext(), funcName))
    results = tmp;
```

Before writing any `empty()` checks, I'd usually check whether `tmp` has `operator bool()` first. In this case `LookupResult` does have one.

dan-zheng replied:
Thanks for the tip! Done.

dan-zheng commented on lib/Sema/TypeCheckExpr.cpp:
```diff
@@ -709,3 +709,107 @@ bool TypeChecker::isCompatibleWithVectorAutoDiff(Type type, DeclContext *DC) {
+  }
+  // If function declaration could not be resolved, emit the appropriate
+  // diagnostic.
+  if (!resolvedFuncDecl) {
```
Done.

dan-zheng committed:
Refactoring/clean up.

Addresses comments from @rxwei.

dan-zheng reviewed (commented):
Thanks! I'll add tests for `@differentiable` access control before merging.

dan-zheng committed:
Add/fix @differentiable attribute diagnostics, clean up.

- Add @differentiable access control check to primal as well as adjoint.
- Add @differentiable access control test.
- Add diagnostic for differentiating void functions.

dan-zheng commented:
@rxwei 
Could you please re-review? I made some slightly non-trivial changes in https://github.com/apple/swift/pull/17246/commits/47f49f6da1880a587e25481d8d9aa48403c4b1e4.

rxwei commented:
LGTM

dan-zheng committed:
Preemptively rename @_versioned to @usableFromInline.

dan-zheng commented:
Good idea about renaming `@_versioned` to `@usableFromInline`, done.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
